compiler/rustc_middle/src/query/mod.rs

```diff
@@ -632,14 +632,21 @@ rustc_queries! {
         return_result_from_ensure_ok
     }
 
-    /// MIR after our optimization passes have run. This is MIR that is ready
-    /// for codegen. This is also the only query that can fetch non-local MIR, at present.
+    /// Polymorphic MIR after our pre-mono optimization passes have run. This is the MIR that
+    /// crates export.
     query optimized_mir(key: DefId) -> &'tcx mir::Body<'tcx> {
         desc { |tcx| "optimizing MIR for `{}`", tcx.def_path_str(key) }
         cache_on_disk_if { key.is_local() }
         separate_provide_extern
     }
 
+    /// MIR for a specific Instance ready for codegen. This is `optimized_mir` but monomorphized
+    /// and with extra transforms applied.
+    query build_codegen_mir(key: ty::Instance<'tcx>) -> &'tcx mir::Body<'tcx> {
+        desc { |tcx| "finalizing codegen MIR for `{}`", tcx.def_path_str_with_args(key.def_id(), key.args) }
+        arena_cache
+    }
```

Comment:
one thing you can try to improve compile times here is `no_hash`. Since we are not storing it in the incremental caches anyway, we need to recompute it always as the only callers are from codegen, which either gets loaded from cache, so we don't need to call this one, or it doesn't get loaded from cache, then we can just reevaluate it. I think `eval_always` would make perf worse again, as we'd rerun it even if it's not needed?